File: client/src/utils/amountInWords.js
const ONES = [
  "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
  "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
  "Seventeen", "Eighteen", "Nineteen",
];
const TENS = [
  "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety",
];

// Below 1000 — the one place ones/tens/hundreds combine before switching to
// the lakh/crore grouping used for everything above it.
function belowThousand(n) {
  if (n === 0) return "";
  if (n < 20) return ONES[n];
  if (n < 100) return TENS[Math.floor(n / 10)] + (n % 10 ? " " + ONES[n % 10] : "");
  return ONES[Math.floor(n / 100)] + " Hundred" + (n % 100 ? " " + belowThousand(n % 100) : "");
}

/**
 * Spells out a rupee amount the way a Nepali/Indian bill does — grouped by
 * thousand/lakh/crore (not the Western million/billion split), e.g.
 * 1,234,567 → "Twelve Lakh Thirty Four Thousand Five Hundred Sixty Seven".
 * Rounds to the nearest rupee; paisa isn't printed on a bill.
 */
export function amountInWords(amount, { currency = "Rupees", onlySuffix = "Only" } = {}) {
  const n = Math.round(Math.abs(Number(amount) || 0));
  if (n === 0) return `Zero ${currency} ${onlySuffix}`;

  const crore = Math.floor(n / 10000000);
  const lakh = Math.floor((n % 10000000) / 100000);
  const thousand = Math.floor((n % 100000) / 1000);
  const rest = n % 1000;

  const parts = [];
  if (crore) parts.push(`${belowThousand(crore)} Crore`);
  if (lakh) parts.push(`${belowThousand(lakh)} Lakh`);
  if (thousand) parts.push(`${belowThousand(thousand)} Thousand`);
  if (rest) parts.push(belowThousand(rest));

  return `${parts.join(" ")} ${currency} ${onlySuffix}`.trim();
}

File: app/lib/core/utils/amount_in_words.dart
const _ones = [
  '', 'One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven', 'Eight', 'Nine',
  'Ten', 'Eleven', 'Twelve', 'Thirteen', 'Fourteen', 'Fifteen', 'Sixteen',
  'Seventeen', 'Eighteen', 'Nineteen',
];
const _tens = [
  '', '', 'Twenty', 'Thirty', 'Forty', 'Fifty', 'Sixty', 'Seventy', 'Eighty', 'Ninety',
];

String _belowThousand(int n) {
  if (n == 0) return '';
  if (n < 20) return _ones[n];
  if (n < 100) {
    return _tens[n ~/ 10] + (n % 10 != 0 ? ' ${_ones[n % 10]}' : '');
  }
  final rest = n % 100;
  return '${_ones[n ~/ 100]} Hundred${rest != 0 ? ' ${_belowThousand(rest)}' : ''}';
}

/// Spells out a rupee amount the Nepali/Indian way — grouped by
/// thousand/lakh/crore, not the Western million/billion split. Mirrors
/// client/src/utils/amountInWords.js so both platforms' bills read
/// identically. Rounds to the nearest rupee; paisa isn't printed on a bill.
String amountInWords(num amount, {String currency = 'Rupees', String onlySuffix = 'Only'}) {
  final n = amount.abs().round();
  if (n == 0) return 'Zero $currency $onlySuffix';

  final crore = n ~/ 10000000;
  final lakh = (n % 10000000) ~/ 100000;
  final thousand = (n % 100000) ~/ 1000;
  final rest = n % 1000;

  final parts = <String>[];
  if (crore != 0) parts.add('${_belowThousand(crore)} Crore');
  if (lakh != 0) parts.add('${_belowThousand(lakh)} Lakh');
  if (thousand != 0) parts.add('${_belowThousand(thousand)} Thousand');
  if (rest != 0) parts.add(_belowThousand(rest));

  return '${parts.join(' ')} $currency $onlySuffix'.trim();
}
